javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -81,83 +85,118 @@ private static int treeSize(int branchCount, int depthShift) {
 
     private BitMappedTrie<T> boxed() { return map(identity()); }
 
-    BitMappedTrie<T> prepend(T leading) {
-        if (cannotAdd(leading)) {
-            return boxed().prepend(leading);
-        } else {
-            final int newSize = length() + 1;
-            if (length() == 0) {
-                return new BitMappedTrie<>(type, type.asArray(leading), offset, newSize, depthShift);
-            } else {
-                Object array = this.array;
-                int shift = depthShift, offset = this.offset;
-                if (isFullLeft()) {
-                    array = obj().copyUpdate(obj().empty(), BRANCHING_FACTOR - 1, array);
-                    shift += BRANCHING_BASE;
-                    offset = treeSize(BRANCHING_FACTOR - 1, shift);
-                }
+    BitMappedTrie<T> prependAll(Iterable<? extends T> iterable) {
+        if (!isTraversableAgain(iterable)) {
+            iterable = Vector(iterable);
```

Comment:
I tried my best, but if we want optimistic modification, this is our best shot, I think.
Made it traversable-again so that we can start from scratch if we can't add the element (will get a `ClassCastException`); box it and try again.